File: Remote/Source/Remote Elements/Commands/Command Subclasses/SendIRCommand.m
//
// SendIR.m
// Remote
//
// Created by Jason Cardwell on 6/1/11.
// Copyright (c) 2011 Moondeer Studios. All rights reserved.
//
#import "Command_Private.h"
#import "IRCode.h"
#import "ComponentDevice.h"

static int ddLogLevel = LOG_LEVEL_DEBUG;
static int msLogContext = (LOG_CONTEXT_COMMAND|LOG_CONTEXT_FILE|LOG_CONTEXT_CONSOLE);
#pragma unused(ddLogLevel,msLogContext)

@implementation SendIRCommand

@dynamic code, portOverride;

- (void)updateCachedCode
{
    IRCode * code = self.code;
    if (code)
    {
        __port        = (self.portOverride ? _portOverride : code.device.port);
        __offset      = code.offset;
        __frequency   = code.frequency;
        __repeatCount = code.repeatCount;
        __pattern     = code.onOffPattern;
        __name        = code.name;
    }
}

- (void)clearCachedCode
{
    __port        = self.portOverride;
    __offset      = 0;
    __frequency   = 0;
    __repeatCount = 0;
    __pattern     = nil;
    __name        = nil;
}

- (void)setCode:(IRCode *)code
{
    [self willChangeValueForKey:@"code"];
    self.primitiveCode = code;
    [self didChangeValueForKey:@"code"];
    [self updateCachedCode];
}

- (void)awakeFromFetch
{
    [super awakeFromFetch];
    if (self.code) [self updateCachedCode];
    else [self clearCachedCode];
}

- (void)setPortOverride:(int16_t)portOverride
{
    [self willChangeValueForKey:@"portOverride"];
    _portOverride = portOverride;
    [self didChangeValueForKey:@"portOverride"];
    __port = (_portOverride  ? : (self.code ? self.code.device.port : 0));
}

- (int16_t)portOverride
{
    [self willAccessValueForKey:@"portOverride"];
    int16_t portOverride = _portOverride;
    [self didAccessValueForKey:@"portOverride"];
    return portOverride;
}

- (ComponentDevice *)device { return self.primitiveCode.device; }

- (int16_t)port { return __port; }

- (int16_t)offset { return __offset; }

- (int16_t)repeatCount { return __repeatCount; }

- (int64_t)frequency { return __frequency; }

- (NSString *)pattern { return __pattern; }

- (NSString *)name { return __name; }

- (NSString *)commandString
{
    return $(@"sendir,1:%i,<tag>,%lld,%i,%i,%@\r",
             __port,
             __frequency,
             __repeatCount,
             __offset,
             __pattern);
}

+ (SendIRCommand *)commandWithIRCode:(IRCode *)code
{
    SendIRCommand * sendIR = [self commandInContext:code.managedObjectContext];
    sendIR.code = code;
    return sendIR;
}

- (NSString *)shortDescription {return $(@"SendIRCommand(%@)", __name);}


////////////////////////////////////////////////////////////////////////////////
#pragma mark Importing
////////////////////////////////////////////////////////////////////////////////


- (void)updateWithData:(NSDictionary *)data {
    /*
         {
             "class": "sendir",
             "code.uuid": "A32C02D7-6CE0-46C0-8469-8F074C6D96E5" // Tools
         }
     */

    [super updateWithData:data];

    NSDictionary * code = data[@"code"];
    if (code) self.code = [IRCode importObjectFromData:code context:self.managedObjectContext];

}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Exporting
////////////////////////////////////////////////////////////////////////////////


- (MSDictionary *)JSONDictionary
{
    MSDictionary * dictionary = [super JSONDictionary];


    dictionary[@"code.uuid"] = CollectionSafe(self.code.commentedUUID);
    dictionary[@"portOverride"] = (self.portOverride ? @(self.portOverride) : NullObject);

    [dictionary compact];
    [dictionary compress];

    return dictionary;
}

@end

File: Remote/Source/Remote Elements/Commands/Command Subclasses/SendIRCommand.swift
//
//  SendIRCommand.swift
//  Remote
//
//  Created by Jason Cardwell on 3/2/15.
//  Copyright (c) 2015 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

/**
  `SendIRCommand` subclasses `Command` to send IR commands via <ConnectionManager> to networked
  IR receivers that control the user's home theater system. At this time, only
  [iTach](http://www.globalcache.com/products/itach) devices from Global Caché are supported.
*/
@objc(SendIRCommand)
class SendIRCommand: SendCommand {

  @NSManaged var portOverride: NSNumber

  @NSManaged var code: IRCode

  var device: ComponentDevice { return code.device }
  var networkDevice: NetworkDevice? { return device.networkDevice }

  var commandString: String {
    return "sendir,1:\(device.port),<tag>,\(code.frequency),\(code.repeatCount),\(code.offset),\(code.onOffPattern)"
  }

  /**
  updateWithData:

  :param: data [String:AnyObject]
  */
  override func updateWithData(data: [String:AnyObject]) {
    super.updateWithData(data)
    if let codeData = data["code"] as? [String:AnyObject], let moc = managedObjectContext,
      let code = IRCode.fetchOrImportObjectWithData(codeData, context: moc) {
        self.code = code
    }
  }

  /**
  JSONDictionary

  :returns: MSDictionary!
  */
  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()

    dictionary["class"] = "sendir"
    dictionary["code.uuid"] = code.uuid
    setIfNotDefault("portOverride", forKey: "port-override", inDictionary: dictionary)

    dictionary.compact()
    dictionary.compress()

    return dictionary
  }

}